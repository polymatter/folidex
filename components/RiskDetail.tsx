import { Ionicons } from "@expo/vector-icons";
import React, { useState, useContext } from "react";
import { Text, View, Switch, TextInput } from '../components/Themed';
import Colors from "../constants/Colors";
import { updateRiskLabel } from "../adaptors/DataAccess";
import Risk from "../entities/Risk";
import RiskStoreContext from "../store/RiskStore";
import RiskLevelBadge from "./RiskLevelBadge";
import { StyleSheet } from 'react-native';

const doNothing = () => { }

const debounce = (func: any, timeout = 300) => {
  let timer: NodeJS.Timeout;
  return (...args: any[]) => {
    clearTimeout(timer);
    timer = setTimeout(() => { func.apply(null, args); }, timeout);
  };
}

const RiskDetail = ({ riskId }: { riskId: string }) => {

  const [editable, setEditable] = useState(false);
  const toggleEditable = () => setEditable(previousState => !previousState);

  const risk: Risk = useContext(RiskStoreContext).filter(r => r.id == riskId)[0]

  if (risk === undefined) {
    return <Text>Undefined</Text>
  }

  const changeLabelHandler = (label: string) => {
    debounce(updateRiskLabel)({ id: risk.id, label });
  }

  return (
    <View>
      <RiskLevelBadge level={risk.level} />
      <View style={styles.test}>
        <TextInput
          value={risk.label}
          onChangeText={editable ? changeLabelHandler : doNothing}
          multiline={true}
          numberOfLines={5}
          returnKeyLabel="done" />
      </View>
      <View style={styles.separator} lightColor="#eee" darkColor="rgba(255,255,255,1)" />
      <View style={styles.bar}>
        <View style={styles.bar}>
          <Ionicons name="people" size={24} color="black" />
          <Text>Shared</Text>
        </View>
        <View style={styles.bar}>
          {editable ?
            <>
              <Ionicons name="lock-open" size={24} color="black" />
              <Text>Editable</Text>
            </>
            :
            <>
              <Ionicons name="lock-closed" size={24} color="black" />
              <Text>Write Protected</Text>
            </>
          }

          <Switch
            onValueChange={toggleEditable}
            value={editable}
          />
        </View>
      </View>
      <View style={styles.separator} lightColor="#eee" darkColor="rgba(255,255,255,1)" />
      <View style={styles.bar}>
        <View style={[styles.bar, styles.sectionOne]}>
          <Text>Mitigation</Text>
        </View>
        <View style={[styles.bar, styles.sectionTwo]}>
          <Text>Contingency</Text>
        </View>
        <View style={[styles.bar, styles.sectionThree]}>
          <Text>Impact</Text>
        </View>
      </View>
      <View style={styles.separator} lightColor="#eee" darkColor="rgba(255,255,255,1)" />
      <View style={styles.section}>
        <Text style={styles.sectionOne}>Mitigation</Text>
        <TextInput value={risk.mitigation} multiline={true} numberOfLines={5} />
      </View>
      <View style={styles.section}>
        <Text style={styles.sectionTwo}>Contingency</Text>
        <TextInput value={risk.contingency} multiline={true} numberOfLines={5} />
      </View>
      <View style={styles.section}>
        <Text style={styles.sectionThree}>Impact</Text>
        <TextInput value={risk.impact} multiline={true} numberOfLines={5} />
      </View>
      <View style={styles.separator} lightColor="#eee" darkColor="rgba(255,255,255,1)" />
    </View>
  )
}

const styles = StyleSheet.create({
  test: {
    paddingLeft: 30,
  },
  bar: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    paddingLeft: 5,
    paddingRight: 5,
    lineHeight: 10,
    minHeight: 30,
    // maxWidth: 310,
  },
  section: {
    padding: 10,
  },
  sectionOne: {
    backgroundColor: Colors.sectionColors[0],
    borderRadius: 20,
    paddingLeft: 15,
  },
  sectionTwo: {
    backgroundColor: Colors.sectionColors[1],
    borderRadius: 20,
    paddingLeft: 15,
  },
  sectionThree: {
    backgroundColor: Colors.sectionColors[0],
    borderRadius: 20,
    paddingLeft: 15,
  },
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
  },
  separator: {
    height: 10,
    width: '100%',
  },
});

export default RiskDetail;