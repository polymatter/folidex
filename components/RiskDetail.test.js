import * as React from 'react';
import { render } from '@testing-library/react-native';
import RiskDetail from './RiskDetail';
import RiskStoreContext from '../store/RiskStore';
import mockdata from '../mocks/RiskStore.json';

describe('<RiskDetail>', () => {

  test('can render a risk detail', () => {

    const tree = render(
      <RiskStoreContext.Provider value={mockdata}>
        <RiskDetail riskId={"Id_1"} showIcon={false} />
      </RiskStoreContext.Provider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  })
})