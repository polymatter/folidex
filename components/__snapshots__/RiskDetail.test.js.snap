// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RiskDetail> can render a risk detail 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "lineHeight": 10,
          "minHeight": 30,
          "paddingLeft": 5,
          "paddingRight": 5,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
          },
          Array [
            Object {
              "borderRadius": 15,
              "display": "flex",
              "flexShrink": 0,
              "fontSize": 12,
              "marginRight": 30,
              "paddingLeft": 8,
              "paddingRight": 8,
            },
            Object {
              "color": "#dc3545",
            },
          ],
        ]
      }
    >
      High
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "paddingLeft": 30,
        },
      ]
    }
  >
    <TextInput
      multiline={true}
      numberOfLines={5}
      onChangeText={[Function]}
      returnKeyLabel="done"
      value="The server capacity initially defined may be inadequate"
    />
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#eee",
        },
        Object {
          "height": 10,
          "width": "100%",
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "lineHeight": 10,
          "minHeight": 30,
          "paddingLeft": 5,
          "paddingRight": 5,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
          },
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "lineHeight": 10,
            "minHeight": 30,
            "paddingLeft": 5,
            "paddingRight": 5,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000",
            },
            undefined,
          ]
        }
      >
        Shared
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
          },
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "lineHeight": 10,
            "minHeight": 30,
            "paddingLeft": 5,
            "paddingRight": 5,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000",
            },
            undefined,
          ]
        }
      >
        Write Protected
      </Text>
      <RCTSwitch
        accessibilityRole="switch"
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "height": 31,
            "width": 51,
          }
        }
        value={false}
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#eee",
        },
        Object {
          "height": 10,
          "width": "100%",
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "lineHeight": 10,
          "minHeight": 30,
          "paddingLeft": 5,
          "paddingRight": 5,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
          },
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "lineHeight": 10,
              "minHeight": 30,
              "paddingLeft": 5,
              "paddingRight": 5,
            },
            Object {
              "backgroundColor": "#DEB789",
              "borderRadius": 20,
              "paddingLeft": 15,
            },
          ],
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000",
            },
            undefined,
          ]
        }
      >
        Mitigation
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
          },
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "lineHeight": 10,
              "minHeight": 30,
              "paddingLeft": 5,
              "paddingRight": 5,
            },
            Object {
              "backgroundColor": "#AFC09D",
              "borderRadius": 20,
              "paddingLeft": 15,
            },
          ],
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000",
            },
            undefined,
          ]
        }
      >
        Contingency
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
          },
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "lineHeight": 10,
              "minHeight": 30,
              "paddingLeft": 5,
              "paddingRight": 5,
            },
            Object {
              "backgroundColor": "#DEB789",
              "borderRadius": 20,
              "paddingLeft": 15,
            },
          ],
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000",
            },
            undefined,
          ]
        }
      >
        Impact
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#eee",
        },
        Object {
          "height": 10,
          "width": "100%",
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "padding": 10,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
          },
          Object {
            "backgroundColor": "#DEB789",
            "borderRadius": 20,
            "paddingLeft": 15,
          },
        ]
      }
    >
      Mitigation
    </Text>
    <TextInput
      multiline={true}
      numberOfLines={5}
      value="Capacity analysis will be done during the design stage, if this shows a problem design issues will be revisited"
    />
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "padding": 10,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
          },
          Object {
            "backgroundColor": "#AFC09D",
            "borderRadius": 20,
            "paddingLeft": 15,
          },
        ]
      }
    >
      Contingency
    </Text>
    <TextInput
      multiline={true}
      numberOfLines={5}
      value="Purchase and install additional disc space at the customer site"
    />
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        Object {
          "padding": 10,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
          },
          Object {
            "backgroundColor": "#DEB789",
            "borderRadius": 20,
            "paddingLeft": 15,
          },
        ]
      }
    >
      Impact
    </Text>
    <TextInput
      multiline={true}
      numberOfLines={5}
      value="Cost of disc space plus installation effort &amp; travel"
    />
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#eee",
        },
        Object {
          "height": 10,
          "width": "100%",
        },
      ]
    }
  />
</View>
`;
